File: webmvp/src/lib/autoscrollSpeed.ts
export const AUTOSCROLL_MIN_SPEED = 0.1;
export const AUTOSCROLL_MAX_SPEED = 2;
export const AUTOSCROLL_DEFAULT_DESKTOP = 0.7;
export const AUTOSCROLL_DEFAULT_TOUCH = 0.4;
export const AUTOSCROLL_TOUCH_RATE_MULTIPLIER = 0.85;
export const AUTOSCROLL_BASE_PX_PER_SEC = 20;

const SLOW_STEP = 0.05;
const FAST_STEP = 0.1;
const STEP_THRESHOLD = 1;

export function clampAutoscrollSpeed(speed: number): number {
  return Math.min(
    AUTOSCROLL_MAX_SPEED,
    Math.max(AUTOSCROLL_MIN_SPEED, Number(speed.toFixed(2))),
  );
}

export function getAutoscrollSpeedStep(speed: number): number {
  return speed < STEP_THRESHOLD ? SLOW_STEP : FAST_STEP;
}

export function decreaseAutoscrollSpeed(current: number): number {
  return clampAutoscrollSpeed(current - getAutoscrollSpeedStep(current));
}

export function increaseAutoscrollSpeed(current: number): number {
  return clampAutoscrollSpeed(current + getAutoscrollSpeedStep(current));
}

export function resolveDefaultAutoscrollSpeed(isTouchDevice: boolean): number {
  return isTouchDevice
    ? AUTOSCROLL_DEFAULT_TOUCH
    : AUTOSCROLL_DEFAULT_DESKTOP;
}

export function resolveAutoscrollPixelsPerSecond(
  speed: number,
  isTouchDevice: boolean,
): number {
  const multiplier = isTouchDevice ? AUTOSCROLL_TOUCH_RATE_MULTIPLIER : 1;
  return AUTOSCROLL_BASE_PX_PER_SEC * speed * multiplier;
}

export function formatAutoscrollSpeed(speed: number): string {
  return speed < STEP_THRESHOLD ? speed.toFixed(2) : speed.toFixed(1);
}

/** Client-only: touch-primary devices get slower default scroll rate. */
export function detectTouchAutoscrollDevice(): boolean {
  if (typeof window === "undefined") {
    return false;
  }

  const query = window.matchMedia(
    "(max-width: 1024px), (pointer: coarse), (hover: none)",
  );
  const hasTouch = navigator.maxTouchPoints > 0 || "ontouchstart" in window;
  return query.matches || hasTouch;
}

File: mobile/lib/domain/autoscroll_speed.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';

const double autoscrollMinSpeed = 0.1;
const double autoscrollMaxSpeed = 2;
const double autoscrollDefaultDesktop = 0.7;
const double autoscrollDefaultTouch = 0.4;
const double autoscrollTouchRateMultiplier = 0.85;
const double autoscrollBasePxPerSec = 20;

const double _slowStep = 0.05;
const double _fastStep = 0.1;
const double _stepThreshold = 1;

double clampAutoscrollSpeed(double speed) {
  final clamped = speed.clamp(autoscrollMinSpeed, autoscrollMaxSpeed);
  return double.parse(clamped.toStringAsFixed(2));
}

double getAutoscrollSpeedStep(double speed) {
  return speed < _stepThreshold ? _slowStep : _fastStep;
}

double decreaseAutoscrollSpeed(double current) {
  return clampAutoscrollSpeed(current - getAutoscrollSpeedStep(current));
}

double increaseAutoscrollSpeed(double current) {
  return clampAutoscrollSpeed(current + getAutoscrollSpeedStep(current));
}

double resolveDefaultAutoscrollSpeed(bool isTouchDevice) {
  return isTouchDevice ? autoscrollDefaultTouch : autoscrollDefaultDesktop;
}

double resolveAutoscrollPixelsPerSecond(
  double speed,
  bool isTouchDevice,
) {
  final multiplier = isTouchDevice ? autoscrollTouchRateMultiplier : 1;
  return autoscrollBasePxPerSec * speed * multiplier;
}

String formatAutoscrollSpeed(double speed) {
  return speed < _stepThreshold
      ? speed.toStringAsFixed(2)
      : speed.toStringAsFixed(1);
}

/// Port of web `detectTouchAutoscrollDevice` using Flutter media signals.
bool detectTouchAutoscrollDevice(BuildContext context) {
  final media = MediaQuery.of(context);
  if (media.size.shortestSide <= 1024) {
    return true;
  }
  return defaultTargetPlatform == TargetPlatform.android ||
      defaultTargetPlatform == TargetPlatform.iOS;
}
